packages/pglite-sync/src/index.ts

```diff
@@ -113,123 +156,183 @@ async function createPlugin(
-                  // Split messageAggregator into initial inserts and remaining messages
-                  const initialInserts: InsertChangeMessage[] = []
-                  const remainingMessages: ChangeMessage<any>[] = []
-                  let foundNonInsert = false
-                  for (const message of messageAggregator) {
-                    if (
-                      !foundNonInsert &&
-                      message.headers.operation === 'insert'
-                    ) {
-                      initialInserts.push(message as InsertChangeMessage)
-                    } else {
-                      foundNonInsert = true
-                      remainingMessages.push(message)
-                    }
-                  }
-                  if (initialInserts.length > 0) {
-                    // As `COPY FROM` doesn't trigger a NOTIFY, we pop
-                    // the last insert message and and add it to the be beginning
-                    // of the remaining messages to be applied after the `COPY FROM`
-                    remainingMessages.unshift(initialInserts.pop()!)
-                  }
-                  messageAggregator = remainingMessages
-
-                  // Do the `COPY FROM` with initial inserts
-                  if (initialInserts.length > 0) {
-                    applyMessagesToTableWithCopy({
-                      pg: tx,
-                      table: options.table,
-                      schema: options.schema,
-                      messages: initialInserts as InsertChangeMessage[],
-                      mapColumns: options.mapColumns,
-                      primaryKey: options.primaryKey,
-                      debug,
-                    })
-                    // We don't want to do a `COPY FROM` again after that
-                    doCopy = false
-                  }
-                }
-
-                for (const changeMessage of messageAggregator) {
-                  await applyMessageToTable({
-                    pg: tx,
-                    table: options.table,
-                    schema: options.schema,
-                    message: changeMessage,
-                    mapColumns: options.mapColumns,
-                    primaryKey: options.primaryKey,
-                    debug,
-                  })
-                }
 
+            if (options.commitGranularity === 'operation') {
+              // commit after each operation if granularity is set to operation
+              await throttledCommit()
+            } else if (typeof options.commitGranularity === 'number') {
+              // commit after every N messages if granularity is set to a number
+              if (messageAggregator.length >= options.commitGranularity) {
+                await throttledCommit()
+              }
+            }
+          } else if (isControlMessage(message)) {
+            switch (message.headers.control) {
+              case 'must-refetch':
```

Comment:
should `onInitialSyncCalled` be reset here? Or do we consider the rebuilding of data an internal thing and thus don't expose any "initial data" info?

Reply:
Yes, I consider it an internal thing.

The idea is that the initialSync callback is only ever called once, that way it can be used to turn on a write path, install triggers, etc. that may add overhead.

TBH, must-refetch adds a lot of pain for large datasets. Deleting everything and reinserting, will be slow, particularly if using Postgres triggers for conflict resolution (as I have in Linearlite). This needs more exploration.